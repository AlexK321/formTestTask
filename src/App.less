.wrapper {
  width: 100vw;
  height: 100vh;
  background: #f3f7fa;
  flex-grow: 1;
  margin: 0;
  display: flex;
  font-family: "Whitney-Book";

  .form-wrapper {
    width: 100%;
    max-width: 770px;
    margin: 70px auto;
    display: flex;
    flex-direction: column;
    justify-content: flex-start; 
    align-items: center;
    
    .form-description {
      text-align: center;
      color: #666;
      font-weight: 400; 
      margin-bottom: 60px;
      line-height: 1.3;
    }
  }
}

h1 {
  font-size: 42px;
  font-family: "Whitney-Medium"; 
}

@font-face { 
  font-family: "Whitney-Book"; 
  src: url("../src/assets/fonts/Whitney-Book.ttf"); 
  font-style: normal; 
  font-weight: normal; 
} 

@font-face { 
  font-family: "Whitney-Light"; 
  src: url("../src/assets/fonts/Whitney-Light.ttf"); 
  font-style: normal; 
  font-weight: normal; 
} 

@font-face { 
  font-family: "Whitney-Medium"; 
  src: url("../src/assets/fonts/Whitney-Medium.ttf"); 
  font-style: normal; 
  font-weight: normal; 
} 