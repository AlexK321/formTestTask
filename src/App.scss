@font-face { 
  font-family: "MPLUS-Regular"; 
  src: url("../src/assets//fonts/MPLUSRounded1c-Regular.ttf") ; 
  font-style: normal; 
  font-weight: normal; 
} 

@font-face { 
  font-family: "OpenSans-Regular"; 
  src: url("../src/assets//fonts/OpenSans-Regular.ttf") ; 
  font-style: normal; 
  font-weight: normal; 
} 

@font-face { 
  font-family: "OpenSans-Light"; 
  src: url("../src/assets//fonts/OpenSans-Light.ttf") ; 
  font-style: normal; 
  font-weight: normal; 
} 

.App {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  margin: 0; 
  width: 100vw;
  height: 100vh;
  font-family: "OpenSans-Regular"; 
}

header,
footer {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  flex-grow: 0;
  flex-shrink: 0;
  background-color: rgb(201, 212, 235);
  width: 100%;
  min-height: 1px;
}
