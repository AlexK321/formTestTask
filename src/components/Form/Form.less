form {
  background-color: white;
  border-radius: 4px;
  padding: 24px 40px!important;
  width: 100%;
  input {
    border-radius: 4px!important;
    padding: 25px 18px 4px 11px!important;
  }
  button {
    border: 1px solid #00b4e3!important;
    background-color: #00b4e3!important;
    border-radius: 8px!important;
    color: #fff!important;
    padding: 0 50px!important;
    height: 60px!important;
    &:hover {
      border-color: #17cfff!important;
      background-color: #17cfff!important;
    }
  }
}  