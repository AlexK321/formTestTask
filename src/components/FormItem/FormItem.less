form {
  background-color: white;
  border-radius: 4px;
  padding: 24px 40px!important;
  width: 100%;
  height: 400px;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;

  .ant-col {
    width: 100%!important;
    max-width: 100%!important;
  }

  .form-row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    p {
      width: 60%;
      margin: auto;
    }
    .ant-row {
      width: 48.5%!important;
      max-width: 100%!important;
    }
  }

  input {
    border-radius: 4px!important;
    padding: 25px 18px 4px 11px!important;
  }

  button {
    border: 1px solid #00b4e3!important;
    background-color: #00b4e3!important;
    border-radius: 8px!important;
    color: #fff!important;
    padding: 0 50px!important;
    height: 60px!important;
    
    &:hover {
      border-color: #17cfff!important;
      background-color: #17cfff!important;
    }
  }
}  