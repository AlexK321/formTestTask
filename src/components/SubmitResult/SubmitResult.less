.result-block{
  background-color: white;
  border-radius: 4px;
  padding: 24px 40px!important;
  width: 100%;
  min-height: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
}  