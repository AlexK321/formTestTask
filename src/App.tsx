import React, { ReactElement } from 'react';
import './App.css';
import Form from './components/Form/Form';
import 'antd/dist/antd.css';

const App = (): ReactElement => {
  return (
    <div className="wrapper">
      <div className="form-wrapper">
        <h1 className="form-title">Заявка на участие</h1>
        <h2 className="form-description">
          Расскажите о проекте, его ценности для бизнеса, оставьте информацию о компании и
          контакты,чтобы мы могли с вами связаться
        </h2>
        <Form />
      </div>
    </div>
  );
};

export default App;
