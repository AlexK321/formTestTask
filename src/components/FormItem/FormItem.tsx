import React, { FC } from 'react';
import { Form, Input } from 'antd';
import './FormItem.css';
import { Rule } from 'antd/lib/form';

interface Props {
  name?: string;
  placeholder?: string;
  rules?: Rule[]; // remove any
}

const DEFAULT_RULES = [
  { required: true, message: 'Поле не должно быть пустым' },
  {
    max: 10,
    message: `Максимальная длинна строки - 10 символов`,
  },
];

//webstorm - Ctr+alt+O - vscode analogs

const FormItem: FC<Props> = ({ name, placeholder, rules = DEFAULT_RULES }) => (
  <Form.Item name={name} rules={rules}>
    <Input name={name} placeholder={placeholder} />
  </Form.Item>
);

export default FormItem;
