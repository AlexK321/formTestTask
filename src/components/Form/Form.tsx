import React, { ReactElement, useState } from 'react';
import { Form as FormComponent, Button } from 'antd';
import './Form.css';
import axios from 'axios';
import SubmitResult from '../SubmitResult/SubmitResult';
import FormItem from '../FormItem/FormItem';

interface FormData {
  email?: string;
  phoneNumber?: string;
  userName?: string;
}

const userName = 'userName';
const phoneNumber = 'phoneNumber';
const email = 'email';
const labelCol = { span: 8 };
const wrapperCol = { span: 16 };

const initialValue = {
  userName: localStorage.getItem(userName) || '',
  phoneNumber: localStorage.getItem(phoneNumber) || '',
  email: localStorage.getItem(email) || '',
};

const changeInitialValue = (initialValue1: FormData, filled1: string | null) => {
  const initialValueArr = Object.entries(initialValue1);
  for (let i = Number(filled1); i < initialValueArr.length; i += 1) {
    initialValueArr[i][1] = '';
  }
  return Object.fromEntries(initialValueArr);
};

const getInitialValuesForm = () => {
  const url = new URL(window.location.href);
  const filled = url.searchParams.get('filled');
  if (filled === '0') {
    url.searchParams.delete('filled');
    document.location.href = url.toString();
  } else if (Number(filled) > 0) {
    return changeInitialValue(initialValue, filled);
  }
  return initialValue;
};

const Form = (): ReactElement => {
  const [requestStatus, setRequestStatus] = useState<number | null>(null);

  const sendFormData = async (formData: FormData) => {
    try {
      const response = await axios.post('https://jsonplaceholder.typicode.com/posts', formData);
      setRequestStatus(response.status);
    } catch (err) {
      console.error(err);
    }
  };

  const handleChange = (formData: FormData): void => {
    const formDataArray = Object.entries(formData);
    localStorage.setItem(String(formDataArray[0][0]), String(formDataArray[0][1]));
  };

  if (requestStatus) return <SubmitResult status={requestStatus} />;

  return (
    <FormComponent
      labelCol={labelCol}
      wrapperCol={wrapperCol}
      initialValues={getInitialValuesForm()}
      onFinish={sendFormData}
      onValuesChange={handleChange}
    >
      <h2>Контакты</h2>
      <FormItem name="userName" placeholder="Ваше имя" />
      <div className="form-row">
        <FormItem name="phoneNumber" placeholder="Телефон" />
        <FormItem name="email" placeholder="E-mail" />
      </div>
      <div className="form-row">
        <Button type="primary" htmlType="submit">
          Отправить заявку
        </Button>
        <p>
          Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
        </p>
      </div>
    </FormComponent>
  );
};

export default Form;
