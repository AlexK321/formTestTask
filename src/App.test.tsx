import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import Form from './components/Form/Form';
import SubmitResult from './components/SubmitResult/SubmitResult';
import FormItem from './components/FormItem/FormItem';

/* eslint-disable */

describe('test App component', () => {
  const setUp = (props: any) => shallow(<App {...props} />)
  let component: any = null;
  beforeEach(() => {
    component = setUp();
  })

  it('should contain title', () => {
    expect(component.text()).toContain('Заявка');
  });

  it('should contain Form', () => {
    const wrapper = component.find('Form')
    expect(wrapper.length).toBe(1);
  });

  it('snapshot component', () => {
    expect(component).toMatchSnapshot();
  });
})

describe('test Form component', () => {
  const setUp = (props: any) => shallow(<Form {...props} />)
  let component: any = null;
  beforeEach(() => {
    component = setUp();
  })
  const consentDataProcessing = 'Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных'

  it('should contain title', () => {
    expect(component.text()).toContain(consentDataProcessing);
  });

  it('should contain button', () => {
    const wrapper = component.find('Button')
    expect(wrapper.length).toBe(1);
  });

  it('snapshot component', () => {
    expect(component).toMatchSnapshot();
  });
})

describe('test SubmitResult component', () => {
  const setUp = (props: any) => shallow(<SubmitResult {...props} />)
  let component: any = null;
  beforeEach(() => {
    component = setUp();
  })

  it('check existence component', () => {
    expect(component).toBeTruthy();
  });

  it('should contain .result-block', () => {
    const wrapper = component.find('.result-block')
    expect(wrapper.length).toBe(1);
  });

  it('snapshot component', () => {
    expect(component).toMatchSnapshot();
  });
})

describe('test props', () => {
  it('should render negative response', () => {
    const response = 500
    const responseMessage = 'Данные формы не отправлены. Ошибка: 500'
    const component = shallow(<SubmitResult status={response} />)
    const description = component.find('h2');
    expect(description.text()).toBe(responseMessage);
  });

  it('should render positive response', () => {
    const response = 200
    const responseMessage = 'Данные формы отправлены успешно. Статус ответа: 200'
    const component = shallow(<SubmitResult status={response} />)
    const description = component.find('h2');
    expect(description.text()).toBe(responseMessage);
  });

  // it('should render placeholder', () => {
  //   const placeholder = 'Enter you name'
  //   const component = mount(<FormItem name='userName' placeholder={placeholder} />)
  //   const input = component.find('input');
  //   expect(input.name).toContain(placeholder);
  // });

  it('check existence element', () => {
    const placeholder = 'Enter you name'
    const component = mount(<FormItem name='userName' placeholder={placeholder} />)
    const input = component.find('input');
    expect(input).toHaveLength(1)
  });

  // it('instanse', () => {
  //   const component = mount(<Form />)
  //   const instanse = component.instance()
  //   console.log(instanse.state)
  //   console.log(instanse)


  // });
})

describe('click test', () => {
  // it('test click', () => {
  //   const component = mount(<Form />)
  //   const button = component.find('button');
  //   button.simulate('click');
  //   expect(component.text).toContain('.result-block');
  // });

  // it('change test', () => {
  //   const component = mount(<FormItem name='userName' placeholder='placeholder' />)
  //   const input = component.find('input');
  //   input.simulate('change');
  //   expect(component.text).toContain('.result-block');
  // });


})

const sum = (a: any, b: any) => {
  return a + b
}

describe('function test', () => {
  const arr1 = [1, 2, 3, 4, 6, 8, 9]
  const arr2 = [2, 3, 4, 6, 8, 9, 10]
  const arr3 = [3, 5, 7, 10, 14, 17, 19]
  it('test click', () => {
    arr1.forEach((item, index) => {
      expect(sum(item, arr2[index])).toBe(arr3[index]);
    })
  });
})
