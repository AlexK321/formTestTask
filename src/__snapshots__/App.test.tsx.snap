// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test App component snapshot component 1`] = `
<div
  className="wrapper"
>
  <div
    className="form-wrapper"
  >
    <h1
      className="form-title"
    >
      Заявка на участие
    </h1>
    <h2
      className="form-description"
    >
      Расскажите о проекте, его ценности для бизнеса, оставьте информацию о компании и контакты,чтобы мы могли с вами связаться
    </h2>
    <Form />
  </div>
</div>
`;

exports[`test Form component snapshot component 1`] = `
<ForwardRef(InternalForm)
  initialValues={
    Object {
      "email": "",
      "phoneNumber": "",
      "userName": "",
    }
  }
  labelCol={
    Object {
      "span": 8,
    }
  }
  onFinish={[Function]}
  onValuesChange={[Function]}
  wrapperCol={
    Object {
      "span": 16,
    }
  }
>
  <h2>
    Контакты
  </h2>
  <FormItem
    name="userName"
    placeholder="Ваше имя"
  />
  <div
    className="form-row"
  >
    <FormItem
      name="phoneNumber"
      placeholder="Телефон"
    />
    <FormItem
      name="email"
      placeholder="E-mail"
    />
  </div>
  <div
    className="form-row"
  >
    <Button
      htmlType="submit"
      type="primary"
    >
      Отправить заявку
    </Button>
    <p>
      Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
    </p>
  </div>
</ForwardRef(InternalForm)>
`;

exports[`test SubmitResult component snapshot component 1`] = `
<div
  className="result-block"
>
  <h2>
    Данные формы отправлены успешно. Статус ответа: undefined
  </h2>
</div>
`;
